### ISSUE TITLE
Propose to generate match games between LZ-40b and elf1 for training.

### ISSUE THREAD

**Issue by bjiyxo** [gcp/leela-zero#1791]
Now LZ-40b #1681  has about the same strength as elf1 under the same visits. However, after investigating the matches more closely, we found that LZ-40b has an extremely high winning rate when the games ended before 180 moves (110:6 for weight b75, 87:6 for weight 390). On the other hand, elf1 has advantages for games ending after 180 moves (67:164 for weight b75, 52:152 for weight 390). Therefore, we think that LZ-40b may be able to learn from matches against elf1. Furthermore, from my investigation on matches based on my go knowledge, elf1 does perform better in the late stage of games. Therefore, we propose to generate match games between LZ-40b and elf1. We think that LZ-40b can learn more from match games than from self-games of LZ-40b or elf1.

We also check the results of match games with the extra parameter –m 999. We generated 432 games. LZ-40b (c9d70c41) has 218W 214L (118:99 for playing white and 100:115 for playing black. https://drive.google.com/file/d/1xvC41z88UsqyEf0ojAv5iHwOIuY3wrSo/view?usp=sharing). This result is similar to the one without randomness. Therefore, there should be no problem to generate match games with some randomness to avoid overfitting due to games being too similar.

**Comment by estonte** [comment]
Has gcp made it official that we're switching to 40b? Can we agree on a mission statement of some sort like in the readme?

1. A Go program with no human provided knowledge. 
2.Using MCTS (but without Monte Carlo playouts) and a deep residual convolutional neural network stack.
3. This is a fairly faithful reimplementation of the system described in the Alpha Go Zero paper "Mastering the Game of Go without Human Knowledge".
4. For all intents and purposes, it is an open source AlphaGo Zero.

1, 2, and 4 are being fufilled. What is a fairly faithful reimplementation of alphago zero? 20b alphago zero generated 4.9 million self-play games at 1600 visits. 40b alphago zero generated 29 millions self-play games at 1600 visits. Leela zero generated 9 millions self-play games up to 15b at 3200 visits and 1 million games of 20b at 1600 visits with help from elf.

If we can agree that this is a fairly faithful reimplementation so far, we're 19 million self-play games of 40b from alphago zero. Do we want to generate 19 million more self-play games on 40b with 1600 visits?

I think we deviated enough from the alphago zero paper that match games between leelazero and elfv1 is very interesting. On the other hand, I'm also interested on how fast 40b can grow from just self-play games.

**Comment by maxinjapan** [comment]
Interesting finding, @bjiyxo 

If the goal remains to improve only from self-play, then probably the answer is no. 

If the goal is to produce a stronger Go program in a shorter amount of time, I do not see why we should not take advantage of ELF1 (and I personally would love this)!   

PS: I cannot help wondering how much of LZ's poor performance in long games is due to the no-resign games with plenty of useless moves...

**Comment by l1t1** [comment]
zero method is copied successful, next step can try others,  include match and modified weights...

**Comment by roy7** [comment]
ELF has a sharper win/loss evaluation than LZ does. So if ELF thinks it will losing, it will resign earlier (making the game shorter). If LZ thinks it is losing, it won't resign as early since the win rate goes down slower (making the game longer).

I'm unconvinced LZ is somehow bad at "long" games or has issues with end game play. She just doesn't give up as early.

**Comment by bjiyxo** [comment]
Though sharper loss evaluation causes short game, the middle game or endgame move of ELF-v1 is still much stronger than that of 40b from my 9D perspective. I can give you some examples, but that will be a little off-topic. I will share it if someone is really interested.

**Comment by maxinjapan** [comment]
Hi @bjiyxo I would be curious to read about this evaluation (although my skill level is nowhere near 9D, sorry).

**Comment by dayun110** [comment]
I checked some of the shorter games (92 moves) from match of 40B vs ELF V1( 2018-08-24 05:24 b75dafd2 VS d13c4099 177 : 170 (51.01%) http://zero.sjeng.org/match-games/5b7f8888cc3dde4a3e822714. I found that many of these games are similar. (game No. of.326, 296, 289, 287, 281,279, 212, 199, 198, 71, 54,35, 33, 22, 15,7, 2 are quite similar.  This means LZ 40B learnt one  trick and use it beat ELF V1 17 times.

**Comment by l1t1** [comment]
@dayun110 can you trainning 20b the same trick?

**Comment by wonderingabout** [comment]
edit : i just changed my proposal to test it, not including no resign until 200, but instead making leela zero 40b play instead of elf v1 starting from move 100, picking 400 games of elf v1 selfplay match (1600 visits), where winrate is 50+/- 5% at move 100

**Comment by Crazyeight101** [comment]
It would make sense that if the 40b is primarily becoming stronger from exposure to ELFv1 self-play, and if there was a lack of samples from ELFv1 self-play for the later stages of the game, and if the samples from earlier stages of the game don't generalize well to later stages of the game, then the 40b, which seems to generalize over the same data better than smaller networks do, would gain an advantage against ELF in the early game, but if that advantage wasn't enough to reach the resign threshold for ELF, then ELF would be able to overcome that advantage through better late game play.  That would seem to explain the distribution of wins over length of games that we're seeing.

If that were the case, then the solution would be to generate more samples of ELF late game positions for training, right?  In which case, @bjiyxo 's proposal to use 40b v ELFv1 games for training seems to make sense if ELFv1 v 40b games tend to be longer on average than ELFv1 self-play games.

**Comment by bjiyxo** [comment]
I will list two different situations, and ELF knows perfectly while 40b cannot.

This is the AlphaGo Master 50 Self-Play Series Game 9.

![1](https://user-images.githubusercontent.com/30645842/44947087-337d8380-ae3a-11e8-8e00-ece6672a3637.PNG)

The left side black group is dead in move 120, but black can make some annoying things there. If you are interested, please check Redmond's review (https://www.youtube.com/watch?v=G0XJhPvQ9sc#t=49m50s). ELF knows to play the trick in a slightly different order, while 20b/40b doesn't.

ELF-v1
![selection_073](https://user-images.githubusercontent.com/30645842/44951385-715bc580-ae95-11e8-94e8-566b91366a09.png)

40B(c9d70c41)
![selection_075](https://user-images.githubusercontent.com/30645842/44951396-a1a36400-ae95-11e8-9530-2035627bb3ef.png)

20B
![selection_074](https://user-images.githubusercontent.com/30645842/44951399-acf68f80-ae95-11e8-94af-80089b5d1cfe.png)


This game was played on Fox Go.

![2](https://user-images.githubusercontent.com/30645842/44947101-7c353c80-ae3a-11e8-8563-16333a2b7b63.PNG)

At move 117, elf-v1 knows to play E7 to kill the whole black group while 20b/40b doesn't.

ELF-v1
![selection_070](https://user-images.githubusercontent.com/30645842/44951374-2b9efd00-ae95-11e8-9060-dacbb350b2a6.png)

40B(c9d7041)
![selection_071](https://user-images.githubusercontent.com/30645842/44951381-4e311600-ae95-11e8-8b1c-7b49bc2fc6d7.png)

20B(b1c37640)
![selection_072](https://user-images.githubusercontent.com/30645842/44951382-6012b900-ae95-11e8-8c95-d430ae346ee5.png)

PS:
I know that giving some examples does not mean elf is better than 20B/40B logically, However, when I use lizzie to review lots of games, elf can give the answer fast and accurately, while the 20B/40B needs much more time to give the answer or hardly give the answer. Judging which one is better is hard to quantify, so I give some examples for your information.

**Comment by wonderingabout** [comment]
examples are just an illustration of your experience with reviewing, you dont need 100 examples because other people could come up with 100 counter examples, since it is hard to quantify without big data, what matters is the general feel that you get from reviewing with both elf and leela zero

**Comment by maxinjapan** [comment]
@bjiyxo definitely not a 9D or pro level.. these are kind of moves I would make! :S

**Comment by maxinjapan** [comment]
@bjiyxo definitely not a 9D or pro level.. these are kind of moves I would make! :S

**Comment by maxinjapan** [comment]
@bjiyxo definitely not a 9D or pro level.. these are the kind of moves I would make! :S
